Make ReadableStreamDefaultReader a C++ class

youennf opened the pull request:
#### 3e58747e49397a85a72c928148fe2c409d130960
<pre>
Make ReadableStreamDefaultReader a C++ class
<a href="https://rdar.apple.com/137978432">rdar://137978432</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=281511">https://bugs.webkit.org/show_bug.cgi?id=281511</a>

Reviewed by Chris Dumez.

To prepare ground for byte stream implementation update, we move ReadableStreamDefaultReader to C++.
This allows better WebIDL typing and will allow to do the byte stream implementation in C++, which will be better for native sources like fetch.

ReadableStreamDefaultReader will be able to handle both regular streams like today and byte streams via an if check, until we move the whole ReadableStream implementation to C++.
ReadableStream now takes a RefPtr&lt;InternalReadableStream&gt; as it will be null for byte stream sources.

We add a bunch of WebIDL definitions following the specification to also prepare readable byte stream implementation.

Covered by existing tests.

* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/general.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/default-reader.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/default-reader.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/default-reader.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/default-reader.any.worker-expected.txt:
* LayoutTests/streams/readable-stream-byob-reader-expected.txt:
* LayoutTests/streams/readable-stream-byob-reader-worker-expected.txt:
* LayoutTests/streams/readable-stream-getReader-expected.txt:
* LayoutTests/streams/reference-implementation/abstract-ops-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/streams/QueuingStrategy.h: Added.
* Source/WebCore/Modules/streams/QueuingStrategy.idl: Added.
* Source/WebCore/Modules/streams/QueuingStrategySize.h: Added.
* Source/WebCore/Modules/streams/QueuingStrategySize.idl: Added.
* Source/WebCore/Modules/streams/ReadableStream.cpp:
* Source/WebCore/Modules/streams/ReadableStream.h:
(WebCore::ReadableStream::internalReadableStream):
(WebCore::ReadableStream::setAsDisturbed):
(WebCore::ReadableStream::ReadableStream):
(WebCore::ReadableStream::lock): Deleted.
(WebCore::ReadableStream::isLocked const): Deleted.
(WebCore::ReadableStream::isDisturbed const): Deleted.
(WebCore::ReadableStream::cancel): Deleted.
* Source/WebCore/Modules/streams/ReadableStream.idl:
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp: Added.
(WebCore::ReadableStreamDefaultReader::create):
(WebCore::ReadableStreamDefaultReader::ReadableStreamDefaultReader):
(WebCore::ReadableStreamDefaultReader::setup):
(WebCore::ReadableStreamDefaultReader::releaseLock):
(WebCore::ReadableStreamDefaultReader::read):
(WebCore::ReadableStreamDefaultReader::genericRelease):
(WebCore::ReadableStreamDefaultReader::errorReadRequests):
(WebCore::ReadableStreamDefaultReader::genericCancel):
(WebCore::ReadableStreamDefaultReader::closed const):
(WebCore::ReadableStreamDefaultReader::resolveClosedPromise):
(WebCore::ReadableStreamDefaultReader::rejectClosedPromise):
(WebCore::ReadableStreamDefaultReader::onClosedPromiseRejection):
(WebCore::ReadableStreamDefaultReader::onClosedPromiseResolution):
(WebCore::JSReadableStreamDefaultReader::read):
(WebCore::JSReadableStreamDefaultReader::closed const):
(WebCore::JSReadableStreamDefaultReader::cancel):
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.h: Added.
(WebCore::ReadableStreamDefaultReader::internalDefaultReader):
(WebCore::ReadableStreamDefaultReader::getNumReadRequests const):
(WebCore::ReadableStreamDefaultReader::addReadRequest):
(WebCore::ReadableStreamDefaultReader::takeFirstReadRequest):
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.idl:
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(createInternalReadableStreamDefaultReader):
(readableStreamDefaultReaderClosedForBindings):
(readableStreamDefaultReaderReadForBindings):
(readableStreamDefaultReaderReleaseLockForBindings):
(readableStreamDefaultReaderCancelForBindings):
(readableStreamDefaultReaderClosedPromise):
(acquireReadableStreamDefaultReader):
(readableStreamError):
(readableStreamState):
(readableStreamStoredError):
(readableStreamGetReaderForBindings): Deleted.
* Source/WebCore/Modules/streams/ReadableStreamReadResult.h: Added.
* Source/WebCore/Modules/streams/ReadableStreamReadResult.idl: Copied from Source/WebCore/Modules/streams/ReadableStreamDefaultReader.idl.
* Source/WebCore/Modules/streams/ReadableStreamType.h: Added.
* Source/WebCore/Modules/streams/ReadableStreamType.idl: Added.
* Source/WebCore/Modules/streams/UnderlyingSource.h: Added.
* Source/WebCore/Modules/streams/UnderlyingSource.idl: Added.
* Source/WebCore/Modules/streams/UnderlyingSourceCancelCallback.h: Added.
* Source/WebCore/Modules/streams/UnderlyingSourceCancelCallback.idl: Added.
* Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.h: Added.
* Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.idl: Added.
* Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.h: Added.
* Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/InternalReadableStream.cpp:
(WebCore::InternalReadableStream::state const):
(WebCore::InternalReadableStream::storedError const):
(WebCore::InternalReadableStream::getReader): Deleted.
* Source/WebCore/bindings/js/InternalReadableStream.h:
* Source/WebCore/bindings/js/InternalReadableStreamDefaultReader.cpp: Added.
(WebCore::invokeReadableStreamDefaultReaderFunction):
(WebCore::InternalReadableStreamDefaultReader::create):
(WebCore::InternalReadableStreamDefaultReader::releaseLock):
(WebCore::InternalReadableStreamDefaultReader::readForBindings):
(WebCore::InternalReadableStreamDefaultReader::closedForBindings):
(WebCore::InternalReadableStreamDefaultReader::cancelForBindings):
(WebCore::InternalReadableStreamDefaultReader::onClosedPromiseRejection):
(WebCore::InternalReadableStreamDefaultReader::onClosedPromiseResolution):
* Source/WebCore/bindings/js/InternalReadableStreamDefaultReader.h: Copied from Source/WebCore/bindings/js/InternalReadableStream.h.
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSDOMPromise.cpp:
(WebCore::DOMPromise::whenSettled):
* Source/WebCore/bindings/js/JSDOMPromise.h:

Canonical link: <a href="https://commits.webkit.org/300392@main">https://commits.webkit.org/300392@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e3135f7f52eda33b8bc60e8377b7441e2202ec8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122405 "16 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42111 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128996 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a5bdd52d-c73f-44e0-b3fd-fdb6197c91f7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124281 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50704 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93038 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/70fdd911-6e2d-4910-b924-78822c674e66) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125357 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34162 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109603 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73695 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/25cb905c-f53a-4459-b549-94324098ba5c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33150 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27760 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72488 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103775 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27970 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131733 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49346 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37555 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101585 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49720 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105823 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101455 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25729 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46831 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24966 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46110 "Hash 4e3135f7 for PR 35231 does not build (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49204 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54947 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48671 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52021 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50353 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/WebCore/bindings/js/InternalReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,138 @@
+ExceptionOr<Ref<InternalReadableStreamDefaultReader>> InternalReadableStreamDefaultReader::create(JSDOMGlobalObject& globalObject, InternalReadableStream& stream)
+{
+    auto* clientData = static_cast<JSVMClientData*>(globalObject.vm().clientData);
+    auto& privateName = clientData->builtinFunctions().readableStreamInternalsBuiltins().createInternalReadableStreamDefaultReaderPrivateName();
```
Have you considered:
1. Making the internals builtins generator use a `JSC::JSInternalFieldObjectImpl` subclass
2. Having a build step for the javascript parts to convert from `@createInternalReadableStream(a, b, c)` into `@getInternalField(@webcoreBuiltins,  1234)(a, b, c)` (where `1234` would be the generated field ID for the specific builtin function)? 

This would avoid the overhead of creating all those identifiers for private builtins at start and avoid the runtime overhead of calling these getters on the global object from C++. It might also make it easier to lazy load the builtin functions

youennf replied:
Plan is to reduce built-ins usage in streams implementation. For now, I will keep the way built-ins are generated but we might indeed optimise this in the future.

cdumez commented on Source/WebCore/Modules/streams/ReadableStream.cpp:
```diff
@@ -90,24 +126,99 @@ ExceptionOr<Vector<Ref<ReadableStream>>> ReadableStream::tee(bool shouldClone)
+void ReadableStream::setDefaultReader(ReadableStreamDefaultReader* reader)
+{
+    ASSERT(!m_defaultReader || !reader);
+    m_defaultReader = WeakPtr { reader };
```
We should not need `WeakPtr { }` here?

youennf replied:
ok

cdumez commented on Source/WebCore/Modules/streams/ReadableStream.h:
```diff
@@ -28,37 +28,78 @@
+    void setDefaultReader(ReadableStreamDefaultReader*);
+    ReadableStreamDefaultReader* defaultReader();
+    
+    void setAsDisturbed() { m_disturbed = true; }
```
nit: `markAsDisturbed()` would sound nicer to me.

youennf replied:
ok

cdumez commented on Source/WebCore/Modules/streams/ReadableStream.h:
```diff
@@ -28,37 +28,78 @@
+    
+    JSC::JSValue storedError(JSDOMGlobalObject&) const;
+    
+    size_t getNumReadRequests() const;
```
I understand they use this naming in the spec but this naming doesn't really follow WebKit style. Ideally, we'd name this `readRequestCount()`

youennf replied:
ok

cdumez commented on Source/WebCore/Modules/streams/ReadableStream.h:
```diff
@@ -28,37 +28,78 @@
+    
 private:
-    const Ref<InternalReadableStream> m_internalReadableStream;
+    bool m_disturbed { false };
```
I would move the smaller types (bool, State) to be the last data members. Size may not matter for this object but it is still good practice.

youennf replied:
I used the order of the spec so that slots are altogether.
I'll keep it like this and we can always optimise this at the end of the implementation effort.

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,251 @@
+void ReadableStreamDefaultReader::read(JSDOMGlobalObject&, Ref<DeferredPromise>&& readRequest)
+{
+    // FIXME: Support reading of byte sources.
+    RefPtr stream = m_stream;
```
nit: Doesn't seem terribly useful at the moment.

youennf replied:
Removed from now, I'll add it back later.

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,251 @@
+// https://streams.spec.whatwg.org/#default-reader-release-lock
+ExceptionOr<void> ReadableStreamDefaultReader::releaseLock(JSDOMGlobalObject& globalObject)
+{
+    if (RefPtr internalDefaultReader = this->internalDefaultReader())
```
Darin often suggests shorter names for small scopes like this. E.g.:
```
if (RefPtr reader = internalDefaultReader())
```
Since the scope is small, it is obvious what kind of reader is it from the initialization.

youennf replied:
ok

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    stream->setDefaultReader(nullptr);
+    stream = nullptr;
```
This is useless, unless you meant to null out `m_stream`?

youennf replied:
Yep, fixed.

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,251 @@
+
+    m_closedRejectionCallback = WTFMove(callback);
+    Ref { m_closedPromise }->whenSettled([weakThis = WeakPtr { *this }] {
+        RefPtr protectedThis = weakThis.get();
```
nit: Maybe we can early return right on the next line if `protectedThis` is null? Instead of initializing another variable and then early return?

youennf replied:
done

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,251 @@
+
+    m_closedResolutionCallback = WTFMove(callback);
+    Ref { m_closedPromise }->whenSettled([weakThis = WeakPtr { *this }] {
+        RefPtr protectedThis = weakThis.get();
```
Ditto about early return.

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,251 @@
+    m_closedResolutionCallback = WTFMove(callback);
+    Ref { m_closedPromise }->whenSettled([weakThis = WeakPtr { *this }] {
+        RefPtr protectedThis = weakThis.get();
+        RefPtr closedPromise = protectedThis ? protectedThis->m_closedPromise.ptr() : nullptr;
```
Then we don't need this ternary.

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,251 @@
+    m_closedRejectionCallback = WTFMove(callback);
+    Ref { m_closedPromise }->whenSettled([weakThis = WeakPtr { *this }] {
+        RefPtr protectedThis = weakThis.get();
+        RefPtr closedPromise = protectedThis ? protectedThis->m_closedPromise.ptr() : nullptr;
```
Then we don't need this ternary.

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.h:
```diff
@@ -0,0 +1,85 @@
+class JSDOMGlobalObject;
+class ReadableStream;
+
+class ReadableStreamDefaultReader : public RefCounted<ReadableStreamDefaultReader>, public CanMakeWeakPtr<ReadableStreamDefaultReader> {
```
You can simply subclass a single class: `RefCountedAndCanMakeWeakPtr<ReadableStreamDefaultReader>`

youennf replied:
done

cdumez commented on Source/WebCore/Modules/streams/ReadableStream.h:
```diff
@@ -28,37 +28,78 @@
+
+using ReadableStreamReader = ReadableStreamDefaultReader;
+
+class ReadableStream : public RefCounted<ReadableStream>, public CanMakeWeakPtr<ReadableStream> {
```
You can simply subclass a single class: `RefCountedAndCanMakeWeakPtr<ReadableStream>`

youennf replied:
done

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.h:
```diff
@@ -0,0 +1,85 @@
+
+    ~ReadableStreamDefaultReader();
+
+    DOMPromise& closed() const;
```
Very confusing name for a promise getter. Can we use a better name internally, even if this ends up being the web-exposed name?

youennf replied:
Changed to closedPromise

cdumez commented on Source/WebCore/Modules/streams/ReadableStreamDefaultReader.h:
```diff
@@ -0,0 +1,85 @@
+    ExceptionOr<void> releaseLock(JSDOMGlobalObject&);
+
+    InternalReadableStreamDefaultReader* internalDefaultReader() { return m_internalDefaultReader.get(); }
+    size_t getNumReadRequests() const { return m_readRequests.size(); }
```
Same comment about the naming not following our style.

youennf replied:
For now, I kept it like this. I'll likely change this in a follow-up.

cdumez commented on Source/WebCore/bindings/js/InternalReadableStream.h:
```diff
@@ -49,18 +49,22 @@ class InternalReadableStream final : public DOMGuarded<JSC::JSObject> {
     JSC::JSValue pipeTo(JSC::JSGlobalObject&, JSC::JSValue, JSC::JSValue);
     JSC::JSValue pipeThrough(JSC::JSGlobalObject&, JSC::JSValue, JSC::JSValue);
 
+    enum class Use { Bindings, Private };
```
Can we use `bool` or `uint8_t` as underlying type?

youennf replied:
done

cdumez commented on Source/WebCore/bindings/js/InternalReadableStreamDefaultReader.cpp:
```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
2024?

cdumez commented on Source/WebCore/bindings/js/InternalReadableStreamDefaultReader.h:
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
2024?

youennf replied:
ok

cdumez reviewed (approved):
LGTM with nits.

youennf committed:
Make ReadableStreamDefaultReader a C++ class
rdar://137978432
https://bugs.webkit.org/show_bug.cgi?id=281511

Reviewed by Chris Dumez.

To prepare ground for byte stream implementation update, we move ReadableStreamDefaultReader to C++.
This allows better WebIDL typing and will allow to do the byte stream implementation in C++, which will be better for native sources like fetch.

ReadableStreamDefaultReader will be able to handle both regular streams like today and byte streams via an if check, until we move the whole ReadableStream implementation to C++.
ReadableStream now takes a RefPtr<InternalReadableStream> as it will be null for byte stream sources.

We add a bunch of WebIDL definitions following the specification to also prepare readable byte stream implementation.

Covered by existing tests.

* LayoutTests/imported/w3c/web-platform-tests/streams/readable-byte-streams/general.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/default-reader.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/default-reader.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/default-reader.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/default-reader.any.worker-expected.txt:
* LayoutTests/streams/readable-stream-byob-reader-expected.txt:
* LayoutTests/streams/readable-stream-byob-reader-worker-expected.txt:
* LayoutTests/streams/readable-stream-getReader-expected.txt:
* LayoutTests/streams/reference-implementation/abstract-ops-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/streams/QueuingStrategy.h: Added.
* Source/WebCore/Modules/streams/QueuingStrategy.idl: Added.
* Source/WebCore/Modules/streams/QueuingStrategySize.h: Added.
* Source/WebCore/Modules/streams/QueuingStrategySize.idl: Added.
* Source/WebCore/Modules/streams/ReadableStream.cpp:
* Source/WebCore/Modules/streams/ReadableStream.h:
(WebCore::ReadableStream::internalReadableStream):
(WebCore::ReadableStream::setAsDisturbed):
(WebCore::ReadableStream::ReadableStream):
(WebCore::ReadableStream::lock): Deleted.
(WebCore::ReadableStream::isLocked const): Deleted.
(WebCore::ReadableStream::isDisturbed const): Deleted.
(WebCore::ReadableStream::cancel): Deleted.
* Source/WebCore/Modules/streams/ReadableStream.idl:
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.cpp: Added.
(WebCore::ReadableStreamDefaultReader::create):
(WebCore::ReadableStreamDefaultReader::ReadableStreamDefaultReader):
(WebCore::ReadableStreamDefaultReader::setup):
(WebCore::ReadableStreamDefaultReader::releaseLock):
(WebCore::ReadableStreamDefaultReader::read):
(WebCore::ReadableStreamDefaultReader::genericRelease):
(WebCore::ReadableStreamDefaultReader::errorReadRequests):
(WebCore::ReadableStreamDefaultReader::genericCancel):
(WebCore::ReadableStreamDefaultReader::closed const):
(WebCore::ReadableStreamDefaultReader::resolveClosedPromise):
(WebCore::ReadableStreamDefaultReader::rejectClosedPromise):
(WebCore::ReadableStreamDefaultReader::onClosedPromiseRejection):
(WebCore::ReadableStreamDefaultReader::onClosedPromiseResolution):
(WebCore::JSReadableStreamDefaultReader::read):
(WebCore::JSReadableStreamDefaultReader::closed const):
(WebCore::JSReadableStreamDefaultReader::cancel):
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.h: Added.
(WebCore::ReadableStreamDefaultReader::internalDefaultReader):
(WebCore::ReadableStreamDefaultReader::getNumReadRequests const):
(WebCore::ReadableStreamDefaultReader::addReadRequest):
(WebCore::ReadableStreamDefaultReader::takeFirstReadRequest):
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.idl:
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(createInternalReadableStreamDefaultReader):
(readableStreamDefaultReaderClosedForBindings):
(readableStreamDefaultReaderReadForBindings):
(readableStreamDefaultReaderReleaseLockForBindings):
(readableStreamDefaultReaderCancelForBindings):
(readableStreamDefaultReaderClosedPromise):
(acquireReadableStreamDefaultReader):
(readableStreamError):
(readableStreamState):
(readableStreamStoredError):
(readableStreamGetReaderForBindings): Deleted.
* Source/WebCore/Modules/streams/ReadableStreamReadResult.h: Added.
* Source/WebCore/Modules/streams/ReadableStreamReadResult.idl: Copied from Source/WebCore/Modules/streams/ReadableStreamDefaultReader.idl.
* Source/WebCore/Modules/streams/ReadableStreamType.h: Added.
* Source/WebCore/Modules/streams/ReadableStreamType.idl: Added.
* Source/WebCore/Modules/streams/UnderlyingSource.h: Added.
* Source/WebCore/Modules/streams/UnderlyingSource.idl: Added.
* Source/WebCore/Modules/streams/UnderlyingSourceCancelCallback.h: Added.
* Source/WebCore/Modules/streams/UnderlyingSourceCancelCallback.idl: Added.
* Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.h: Added.
* Source/WebCore/Modules/streams/UnderlyingSourcePullCallback.idl: Added.
* Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.h: Added.
* Source/WebCore/Modules/streams/UnderlyingSourceStartCallback.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/InternalReadableStream.cpp:
(WebCore::InternalReadableStream::state const):
(WebCore::InternalReadableStream::storedError const):
(WebCore::InternalReadableStream::getReader): Deleted.
* Source/WebCore/bindings/js/InternalReadableStream.h:
* Source/WebCore/bindings/js/InternalReadableStreamDefaultReader.cpp: Added.
(WebCore::invokeReadableStreamDefaultReaderFunction):
(WebCore::InternalReadableStreamDefaultReader::create):
(WebCore::InternalReadableStreamDefaultReader::releaseLock):
(WebCore::InternalReadableStreamDefaultReader::readForBindings):
(WebCore::InternalReadableStreamDefaultReader::closedForBindings):
(WebCore::InternalReadableStreamDefaultReader::cancelForBindings):
(WebCore::InternalReadableStreamDefaultReader::onClosedPromiseRejection):
(WebCore::InternalReadableStreamDefaultReader::onClosedPromiseResolution):
* Source/WebCore/bindings/js/InternalReadableStreamDefaultReader.h: Copied from Source/WebCore/bindings/js/InternalReadableStream.h.
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSDOMPromise.cpp:
(WebCore::DOMPromise::whenSettled):
* Source/WebCore/bindings/js/JSDOMPromise.h:

Canonical link: https://commits.webkit.org/300392@main
